pkg/quota/controller/clusterquotamapping/clusterquotamapping.go

```diff
@@ -0,0 +1,536 @@
+	return exists && kapi.Semantic.DeepEqual(labels, namespace.Labels)
+}
+
+func (m *clusterQuotaMapper) setMapping(quota *quotaapi.ClusterResourceQuota, namespace *kapi.Namespace, remove bool) bool {
+	m.lock.RLock()
+	selector, selectorExists := m.quotaToSelector[quota.Name]
+	labels, labelsExist := m.namespaceToLabels[namespace.Name]
+	m.lock.RUnlock()
+
+	if !selectorMatches(selector, selectorExists, quota) {
+		return false
+	}
+	if !labelsMatch(labels, labelsExist, namespace) {
+		return false
+	}
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	selector, selectorExists = m.quotaToSelector[quota.Name]
+	labels, labelsExist = m.namespaceToLabels[namespace.Name]
+	if !selectorMatches(selector, selectorExists, quota) {
+		return false
+	}
+	if !labelsMatch(labels, labelsExist, namespace) {
+		return false
+	}
+
+	if remove {
+		namespaces, ok := m.quotaToNamespaces[quota.Name]
+		if !ok {
+			m.quotaToNamespaces[quota.Name] = sets.String{}
+		} else {
+			namespaces.Delete(namespace.Name)
+		}
+
+		quotas, ok := m.namespaceToQuota[namespace.Name]
+		if !ok {
+			m.namespaceToQuota[namespace.Name] = sets.String{}
+		} else {
+			quotas.Delete(quota.Name)
+		}
+
+		return true
+	}
+
+	namespaces, ok := m.quotaToNamespaces[quota.Name]
+	if !ok {
+		m.quotaToNamespaces[quota.Name] = sets.NewString(namespace.Name)
+	} else {
+		namespaces.Insert(namespace.Name)
+	}
+
+	quotas, ok := m.namespaceToQuota[namespace.Name]
+	if !ok {
+		m.namespaceToQuota[namespace.Name] = sets.NewString(quota.Name)
+	} else {
+		quotas.Insert(quota.Name)
+	}
+
+	return true
+
+}
+
+func (c *ClusterQuotaMappingController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Wait for the stores to sync before starting any work in this controller.
+	ready := make(chan struct{})
+	go c.waitForSyncedStores(ready, stopCh)
+	select {
+	case <-ready:
+	case <-stopCh:
+		return
+	}
+
+	glog.V(4).Infof("Starting workers for quota mapping controller workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.namespaceWorker, time.Second, stopCh)
+		go wait.Until(c.quotaWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down quota mapping controller")
+	c.namespaceQueue.ShutDown()
+	c.quotaQueue.ShutDown()
+}
+
+func (c *ClusterQuotaMappingController) syncQuota(quota *quotaapi.ClusterResourceQuota) error {
+	if !c.clusterQuotaMapper.requireQuota(quota) {
+		return nil
+	}
+
+	selector, err := unversioned.LabelSelectorAsSelector(quota.Spec.Selector)
+	if err != nil {
+		return err
+	}
+
+	allNamespaces, err := c.namespaceLister.List(kapi.ListOptions{})
+	if err != nil {
+		return err
+	}
+	for i := range allNamespaces {
+		namespace := allNamespaces[i]
+
+		// attempt to set the mapping. The quotas never collide with each other (same quota is never processed twice in parallel)
+		// so this means that the project we have is out of date, pull a more recent copy from the cache and retest
+		for {
```

Comment:
This can hotloop.  I think I'll update it so that the `requireQuota` call happens in the "add/update" calls and you re-lookup both resources.  At that point I think both looks can collapse and the "early" half of the loop will be re-calculated by the queued item.